docs/modules/httpd.md

```diff
@@ -0,0 +1,286 @@
+    - `httpd.INDEX_ALL` Any path ending with a "/" has "index.html" appended.
+    For example, a request for "subdir/" would become "subdir/index.html",
+    which in turn might result in the file "web/subdir/index.html" being
+    served (if the `webroot` was set to "web").
+    The default value is `httpd.INDEX_ROOT`.
+
+#### Returns
+`nil`
+
+#### Example
+```lua
+httpd.start({ webroot = "web", auto_index = httpd.INDEX_ALL })
+```
+
+## httpd.stop()
+
+Stops the web server. All registered route handlers are removed.
+
+#### Syntax
+```lua
+httpd.stop()
+```
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+
+## httpd.static()
+
+Registers a static route handler.
+
+#### Syntax
+```
+httpd.static(route, content_type)
+```
+
+#### Parameters
+- `route` The route prefix. Typically in the form of \*.ext to serve all files
+with the ".ext" extension statically. Refer to the Espressif [documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/esp_http_server.html)
+if you wish to use the native Espressif style of static routes instead.
+- `content_type` The value to send in the `Content-Type` header for this file
+type.
+
+#### Returns
+An error code on failure, or `nil` on success. The error code is the value
+returned from the `httpd_register_uri_handler()` function.
+
+#### Example
+```lua
+httpd.start({ webroot = "web" })
+httpd.static("*.csv", "text/csv") -- Serve CSV files under web/
+```
+
+## httpd.dynamic()
+
+Registers a dynamic route handler.
+
+#### Syntax
+```lua
+httpd.dynamic(method, route, handler)
+```
+
+#### Parameters
+- `method` The HTTP method this route applies to. One of:
+  - `httpd.GET`
+  - `httpd.HEAD`
+  - `httpd.PUT`
+  - `httpd.POST`
+  - `httpd.DELETE`
+- `route` The route prefix. Be mindful of any trailing "/" as that may interact
+with the `auto_index` functionality.
+- `handler` The route handler function - `handler(req)`. The provided request
+object `req` has the following fields/functions:
+  - `method` The request method. Same as the `method` parameter above. If the
+  same function is registered for several methods, this field can be used to
+  determine the method the request used.
+  - `uri` The requested URI. Includes both path and query string (if
+  applicable).
+  - `query` The query string on its own. Not decoded.
+  - `headers` A table-like object in which request headers may be looked up.
+  Note that due to the Espressif API not providing a way to iterate over all
+  headers this table will appear empty if fed to `pairs()`.
+  - `getbody()` A function which may be called to read in the request body
+  incrementally. The size of each chunk is set via the Kconfig option
+  "Receive body chunk size". When this function returns `nil` the end of
+  the body has been reached. May raise an error if reading the body fails
+  for some reason (e.g. timeout, network error).
+
+Note that the provided `req` object is _only valid_ within the scope of this
```

Comment:
How does this work with getbody()? Do I have to finish consuming the body in this callback? This might imply that the entire request has been bffered by the espressif server before the callback is invoked.

What actually happens if you do store this object? I hope that it doesn't crash the runtime.....

Reply:
Yes, you have to finish processing the body within the call-back. This is a restriction of the esp_http_server. Afaik what it does underneath the hood is to close the receive window to the sender to avoid having to buffer excessively.

If you were to call `getbody()` (or access the headers object) outside the callback, you'll get a Lua error raised telling you off for doing so. Again, this is due to the limitations of the esp_http_server component itself.